File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        piles.sort()
        lo, hi = 1, piles[-1]

        while lo < hi:
            mid = (lo + hi) // 2
            cnt = 0

            for pile in piles:
                cnt += pile // mid
                if pile % mid:
                    cnt += 1

            if cnt > h:
                lo = mid + 1
            else:
                hi = mid

        return lo

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        Arrays.sort(piles);
        int n = piles.length;
        int lo = 1, hi = piles[n - 1];

        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            int cnt = 0;

            for (int pile: piles) {
                cnt += pile / mid;
                if (pile % mid != 0) {
                    cnt++;
                }
            }

            if (cnt > h) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }

        return lo;
    }
}